components/builder-api/src/server/resources/origins.rs

```diff
@@ -516,46 +516,61 @@ fn upload_origin_key(req: HttpRequest,
                      -> HttpResponse {
     let (origin, revision) = path.into_inner();
 
-    let account_id =
-        match authorize_session(&req, Some(&origin), Some(OriginMemberRole::Administrator)) {
-            Ok(session) => session.get_id(),
-            Err(err) => return err.into(),
-        };
+    // Since this route allows users to upload keys, we verify their membership
+    // before we determine if the key they're using actually exists
+    if let Err(err) = authorize_session(&req, Some(&origin), Some(OriginMemberRole::Member)) {
+        return err.into();
+    }
 
     let conn = match state.db.get_conn().map_err(Error::DbError) {
         Ok(conn_ref) => conn_ref,
         Err(err) => return err.into(),
     };
 
-    match parse_key_str(&body) {
-        Ok((PairType::Public, ..)) => {
-            debug!("Received a valid public key");
-        }
-        Ok(_) => {
-            debug!("Received a secret key instead of a public key");
-            return HttpResponse::new(StatusCode::UNPROCESSABLE_ENTITY);
-        }
-        Err(e) => {
-            debug!("Invalid public key content: {}", e);
-            return HttpResponse::new(StatusCode::UNPROCESSABLE_ENTITY);
-        }
-    }
+    match OriginPublicSigningKey::get(&origin, &revision, &conn) {
+        Ok(_) => HttpResponse::new(StatusCode::CONFLICT),
```

Comment:
Do we have a way to test that the hab pkg upload command behaves as we expect (and will continue to do so?)

Reply:
We don't have a way to automatically test this in our current testing paradigm. We could add something to the pipeline to do so. But I don't think it would be part of this PR. The basic upload behaviors are already thoroughly tested however.